File: promo-video/generate-voiceover.py
"""
Generate voiceover audio for NaijaTransfer promo video using MeloTTS (OpenVoice V2 base).
Produces natural-sounding English speech with no voice overlay between scenes.

Usage: conda run -n openvoice python generate-voiceover.py
"""

import os
import torch
from melo.api import TTS

OUTPUT_DIR = "public/audio"
os.makedirs(OUTPUT_DIR, exist_ok=True)

# MeloTTS supports: EN, ES, FR, ZH, JP, KR
# Use English with American accent for clear, professional voice
device = "cpu"  # Use "cuda" if GPU available
speed = 0.95  # Slightly slower for clarity

# Voiceover scenes — each scene gets its own audio file (no overlay)
scenes = [
    {
        "id": "vo-hook",
        "text": "Tired of WhatsApp destroying your files? Email caps you at just twenty-five megabytes.",
    },
    {
        "id": "vo-problem",
        "text": "Photos lose quality. Videos get crushed. Large files? Forget it. And Nigerian internet makes it worse.",
    },
    {
        "id": "vo-comparison",
        "text": "WhatsApp, Email, WeTransfer. They all fall short. NaijaTransfer is the clear winner.",
    },
    {
        "id": "vo-reveal",
        "text": "Introducing NaijaTransfer.",
    },
    {
        "id": "vo-showcase",
        "text": "Upload up to four gigabytes completely free. No account needed. Resumable uploads built for Nigerian internet.",
    },
    {
        "id": "vo-security",
        "text": "Password protection. WhatsApp sharing. Auto expiry. And your files stay in original quality. Always.",
    },
    {
        "id": "vo-nigeria",
        "text": "Built in Nigeria, for Nigeria. All prices in Naira. Free forever. Pro from just two thousand Naira a month.",
    },
    {
        "id": "vo-cta",
        "text": "Visit naijatransfer dot com. Send files for free. No account. No wahala.",
    },
]

def main():
    print("Loading MeloTTS model...")
    model = TTS(language="EN", device=device)

    # Get available speaker IDs
    speaker_ids = model.hps.data.spk2id
    print(f"Available speakers: {speaker_ids}")

    # Use EN-US (American English) for a clear, professional voice
    # Available: EN-US, EN-BR (British), EN-AU (Australian), EN-Default
    speaker_key = "EN-US"
    if speaker_key not in speaker_ids:
        speaker_key = list(speaker_ids.keys())[0]
    speaker_id = speaker_ids[speaker_key]
    print(f"Using speaker: {speaker_key} (id={speaker_id})")

    for scene in scenes:
        out_path = os.path.join(OUTPUT_DIR, f"{scene['id']}.mp3")
        print(f"Generating: {scene['id']}...")

        # Generate speech — outputs WAV, we convert to MP3 path
        # MeloTTS output_path determines format by extension
        wav_path = out_path.replace(".mp3", ".wav")
        model.tts_to_file(
            scene["text"],
            speaker_id,
            wav_path,
            speed=speed,
        )

        # Convert WAV to MP3 using ffmpeg for smaller file size
        os.system(f'ffmpeg -i "{wav_path}" -codec:a libmp3lame -qscale:a 2 "{out_path}" -y -loglevel quiet')
        os.remove(wav_path)

        file_size = os.path.getsize(out_path) / 1024
        print(f"  Saved: {out_path} ({file_size:.1f} KB)")

    print("\nAll voiceovers generated!")
    print("Now run: cd promo-video && npx remotion render NaijaTransferPromo out/naijatransfer-promo-v2.mp4")

if __name__ == "__main__":
    main()

File: promo-video/generate-voiceover.ts
import { writeFileSync, mkdirSync, existsSync } from "fs";

/**
 * Generate voiceover audio using ElevenLabs TTS API.
 *
 * Usage: node --env-file=.env generate-voiceover.ts
 *
 * Requires ELEVENLABS_API_KEY in .env file.
 */

const API_KEY = process.env.ELEVENLABS_API_KEY;
if (!API_KEY) {
  console.error("Missing ELEVENLABS_API_KEY in .env file");
  process.exit(1);
}

// Use a natural-sounding voice — "Adam" is a warm, clear male voice
// You can browse voices at https://elevenlabs.io/voice-library
const VOICE_ID = "pNInz6obpgDQGcFmaJgB"; // Adam

const scenes = [
  {
    id: "vo-hook",
    text: "Tired of WhatsApp destroying your files? Email caps you at just 25 megabytes.",
  },
  {
    id: "vo-problem",
    text: "Photos lose quality. Videos get crushed. Large files? Forget it. And Nigerian internet makes it worse.",
  },
  {
    id: "vo-comparison",
    text: "WhatsApp, Email, WeTransfer — they all fall short. NaijaTransfer is the clear winner.",
  },
  {
    id: "vo-reveal",
    text: "Introducing NaijaTransfer.",
  },
  {
    id: "vo-showcase",
    text: "Upload up to 4 gigabytes completely free. No account needed. Resumable uploads built for Nigerian internet.",
  },
  {
    id: "vo-security",
    text: "Password protection. WhatsApp sharing. Auto expiry. And your files stay in original quality. Always.",
  },
  {
    id: "vo-nigeria",
    text: "Built in Nigeria, for Nigeria. All prices in Naira. Free forever. Pro from just two thousand Naira a month.",
  },
  {
    id: "vo-cta",
    text: "Visit naijatransfer dot com. Send files for free. No account. No wahala.",
  },
];

const OUTPUT_DIR = "public/audio";

async function generateVoiceover(id: string, text: string) {
  console.log(`Generating: ${id}...`);

  const response = await fetch(
    `https://api.elevenlabs.io/v1/text-to-speech/${VOICE_ID}`,
    {
      method: "POST",
      headers: {
        "xi-api-key": API_KEY!,
        "Content-Type": "application/json",
        Accept: "audio/mpeg",
      },
      body: JSON.stringify({
        text,
        model_id: "eleven_multilingual_v2",
        voice_settings: {
          stability: 0.6,
          similarity_boost: 0.8,
          style: 0.2,
          use_speaker_boost: true,
        },
      }),
    }
  );

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`ElevenLabs API error for ${id}: ${response.status} - ${error}`);
  }

  const audioBuffer = Buffer.from(await response.arrayBuffer());
  const path = `${OUTPUT_DIR}/${id}.mp3`;
  writeFileSync(path, audioBuffer);
  console.log(`  Saved: ${path} (${(audioBuffer.length / 1024).toFixed(1)} KB)`);
}

async function main() {
  if (!existsSync(OUTPUT_DIR)) {
    mkdirSync(OUTPUT_DIR, { recursive: true });
  }

  // Generate sequentially to avoid rate limits and ensure no voice overlap
  for (const scene of scenes) {
    await generateVoiceover(scene.id, scene.text);
  }

  console.log("\nAll voiceovers generated! Now generate background music and SFX:");
  console.log("  - public/audio/bg-music.mp3 (ambient background, ~30s loop)");
  console.log("  - public/audio/sfx-whoosh.mp3 (short whoosh for logo reveal)");
  console.log("\nThen run: npm run render");
}

main().catch(console.error);
